pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -71,35 +75,76 @@ func getDefaultCircuitBreakerThresholds() *cluster.CircuitBreakers_Thresholds {
 // For outbound: Cluster for each service/subset hostname or cidr with SNI set to service hostname
 // Cluster type based on resolution
 // For inbound (sidecar only): Cluster for each inbound endpoint port and for each service port
-func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *model.PushContext) ([]*discovery.Resource, model.XdsLogDetails) {
+// if attemptDeltas is set to true, then BuildClusters will attempt to build delta clusters for that proxy
+func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *model.PushContext, attemptDeltas bool, updates *model.PushRequest, watched *model.WatchedResource) ([]*discovery.Resource, model.XdsLogDetails) {
 	clusters := make([]*cluster.Cluster, 0)
 	resources := model.Resources{}
+	services := make([]*model.Service, 0)
 	envoyFilterPatches := push.EnvoyFilters(proxy)
 	cb := NewClusterBuilder(proxy, push, configgen.Cache)
 	instances := proxy.ServiceInstances
-	cacheStats := cacheStats{}
+	configKeys := configMapKeys(updates.ConfigsUpdated)
+	delta := len(configKeys) == 1 && configKeys[0].Kind == gvk.Service && attemptDeltas
+	if delta {
+		// use delta -- only services changed
+		configs := model.ConfigsOfKind(updates.ConfigsUpdated, gvk.Service)
+		for s := range configs {
+			// get service that changed
+			service := push.ServiceForHostname(proxy, host.Name(nameFromFQDN(s.Name)))
+			// is the service visible to envoy?
+			// does envoy care about this service changing? (watchedResources)
+			if push.IsServiceVisible(service, proxy.ConfigNamespace) {
```

Comment:
I don't think this fully takes
```
	if features.FilterGatewayClusterConfig && cb.proxy.Type == model.Router {
		services = cb.push.GatewayServices(cb.proxy)
	} else {
		services = cb.push.Services(cb.proxy)
	}
```

logic into account.

For gateways, its slightly different. For sidecar, `push.IsServiceVisible` only accounts for if its _exported_, not if its imported. I think we need to look at `proxy.SidecarScope.serviceByHostname`